src/Client/ConnectionEstablisher.h

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include <Common/Epoll.h>
+#include <Common/Fiber.h>
+#include <Common/FiberStack.h>
+#include <Common/TimerDescriptor.h>
+#include <Common/PoolWithFailoverBase.h>
+#include <Client/ConnectionPool.h>
+
+namespace DB
+{
+
+/// Class for nonblocking establishing connection to the replica.
+/// It runs establishing connection process in fiber and sets special
+/// read callback which is called when reading from socket blocks.
+/// When read callback is called, socket and receive timeout are added in epoll
+/// and execution returns to the main program.
+/// So, you can poll this epoll file descriptor to determine when to resume.
+class ConnectionEstablisher
+{
+public:
+    using TryResult = PoolWithFailoverBase<IConnectionPool>::TryResult;
+
+    ConnectionEstablisher(IConnectionPool * pool_,
+                          const ConnectionTimeouts * timeouts_,
+                          const Settings * settings_,
+                          const QualifiedTableName * table_to_check = nullptr);
+
+    /// Establish connection with replica, call async_callbeck when
+    /// reading from socket blocks.
+    void establishConnection(AsyncCallback async_callback = {});
```

Comment:
As my code search shows, there is only one public usage of this method. With empty callback.
I think it would be better to remove `async_callback` from public interface. Just make something like `establishConnectionImpl(AsyncCallback async_callback)` and implement `establishConnection()` as `establishConnectionImpl({})`

Reply:
I think, it is even better to split this interface into two parts. One is just class with `establishConnection` implementation, which may be used externally, without callbacks. And another one with `resume` + `cancel`  interface and callbacks, which reuses first part.